Add Indent Support in to_json

WillAyd opened the pull request:
- [X] closes #12004
- [X] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

After some of the recent refactor this should be relatively easy to do after vendoring a few updates

WillAyd committed:
Added test

WillAyd committed:
Vendored ujson changes

WillAyd committed:
Pass indent to ext module

WillAyd committed:
hacked together working example

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
Expanded test and clean implementation

WillAyd committed:
Cleaned up docs and whitespace

gfyoung commented on pandas/core/generic.py:
```diff
@@ -2329,6 +2330,11 @@ def to_json(
 
             .. versionadded:: 0.23.0
 
+        indent : integer, default 0
```
The parameter name doesn't seem 100% intuitive given the description.  Perhaps `indentLength` ?

WillAyd replied:
Indifferent though the stdlib uses indent as well, so naming consistency with that might have some advantages

simonjayhawkins replied:
in io.formats.html, `indent_delta` is used, although this isn't yet part of the public API.

we should probably be consistent. so if `indent` is used here, would also need to update io.formats.html

gfyoung replied:
> Indifferent though the stdlib uses indent as well

They also use `ensure_ascii`, while we use `force_ascii` 😉

simonjayhawkins replied:
i agree.

simonjayhawkins commented on pandas/core/generic.py:
```diff
@@ -2249,6 +2249,7 @@ def to_json(
         lines=False,
         compression="infer",
         index=True,
+        indent=0,
```
add type hints when modifying code?

WillAyd committed:
Annotated to_json in pandas.core.generic

WillAyd committed:
Simple annotations for to_json

WillAyd committed:
More hints in _json

WillAyd committed:
blackify

WillAyd commented on pandas/core/generic.py:
```diff
@@ -2239,17 +2242,18 @@ def to_excel(
+        double_precision: int = 10,
+        force_ascii: bool_t = True,
+        date_unit: str = "ms",
+        default_handler: Optional[Callable[[Any], Union[int, str, List, Dict]]] = None,
```
The return for the annotation here should match the `Serializable` alias in the _json file. Not sure if better to put in a shared place like _typing or duplicate here; open to suggestions

WillAyd commented on pandas/io/json/_json.py:
```diff
@@ -31,20 +32,23 @@
 
 TABLE_SCHEMA_VERSION = "0.20.0"
 
+Serializable = Union[int, str, List, Dict]
```
If we wanted to go crazy could subscript List and Dict recursively and with int or str, but I don't think mypy yet supports recursive types. Figure this is a good enough tradeoff

WillAyd commented on pandas/io/json/_json.py:
```diff
@@ -31,20 +32,23 @@
 # interface to/from
 def to_json(
     path_or_buf,
     obj,
```
I think `obj` should be FrameOrSeries, but since it gets passed too pandas.core.generic I was getting errors like

```sh
pandas/core/generic.py:2401: error: Value of type variable "FrameOrSeries" of "to_json" cannot be "NDFrame"
```

simonjayhawkins replied:
I think FrameOrSeries in _typing should be changed to `FrameOrSeries = TypeVar("FrameOrSeries", bound="NDFrame")` as used in #27646

WillAyd commented on pandas/io/json/_json.py:
```diff
@@ -97,18 +102,19 @@ class Writer:
 
         if orient is None:
-            orient = self._default_orient
+            orient = self._default_orient  # type: ignore
```
`orient` is probably better as an abstract property but I figure best to leave that refactor to a follow up

simonjayhawkins replied:
add a comment?

although adding an abstract property is straightforward,  if you add if TYPE-CHECKING it won't affect runtime behaviour.

```
if TYPE_CHECKING:
   _default_orient = None  # type: <...>
```

simonjayhawkins replied:
no problem.

I don't think we should avoid these blocks as too beneficial.

in a (hopefully not too distant) future PR, i'll be adding...

```
    if TYPE_CHECKING:  # attributes index and columns are created dynamically
        index = None  # type: Index
        columns = None  # type: Index
```

to frame.py so that i can get types for `self.fmt.tr_frame.index.format` and `self.frame.index.nlevels` etc.

in this case all tests pass without the `if TYPE_CHECKING` but would rather not change runtime behaviour if avoidable.

WillAyd commented on pandas/io/json/_json.py:
```diff
@@ -31,20 +32,23 @@
 
 # interface to/from
 def to_json(
     path_or_buf,
```
I think this should be `FilePathOrBuffer` but was failing with

```sh
pandas/io/json/_json.py:97: error: Item "Path" of "Union[Path, IO[Any]]" has no attribute "write"
```

I think this is actually a problem with the return of `_stringify_path` since it doesn't indicate that Path objects get converted to strings. Can open an issue as a follow up but @simonjayhawkins may already be working on that

simonjayhawkins replied:
yeah, should already be fixed with overloads in #27732.

WillAyd reviewed (commented):
OK type hints added where feasible, comments where I think not

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

simonjayhawkins commented:
@WillAyd can you merge master.

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
Reused Scalar variable

WillAyd committed:
Fixed vendored changes

WillAyd committed:
Replaced tabs with spaces

WillAyd committed:
isort fixup

WillAyd committed:
whatsnew

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd commented:
For completeness here are benchmarks showing no change:

<details>

```sh
· Running 22 total benchmarks (2 commits * 1 environments * 11 benchmarks)
[  0.00%] · For pandas commit 51db82d9 <master> (round 1/2):
[  0.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...............................................
[  0.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[  6.82%] ··· Running (io.json.ToJSON.time_to_json--).
[  9.09%] ··· Running (io.json.ToJSON.time_to_json_wide--).
[ 11.36%] ··· Running (io.json.ToJSONLines.time_delta_int_tstamp_lines--).....
[ 25.00%] · For pandas commit 638d0552 <json-indent2> (round 1/2):
[ 25.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 25.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 31.82%] ··· Running (io.json.ToJSON.time_to_json--).
[ 34.09%] ··· Running (io.json.ToJSON.time_to_json_wide--).
[ 36.36%] ··· Running (io.json.ToJSONLines.time_delta_int_tstamp_lines--).....
[ 50.00%] · For pandas commit 638d0552 <json-indent2> (round 2/2):
[ 50.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 52.27%] ··· io.json.ToJSON.mem_to_json                                                                                    ok
[ 52.27%] ··· ========= ==== ============= ============== =============== ==================
              --                                       frame
              --------- --------------------------------------------------------------------
                orient   df   df_date_idx   df_td_int_ts   df_int_floats   df_int_float_str
              ========= ==== ============= ============== =============== ==================
                split    0         0             0               0                0
               columns   0         0             0               0                0
                index    0         0             0               0                0
                values   0         0             0               0                0
               records   0         0             0               0                0
              ========= ==== ============= ============== =============== ==================

[ 54.55%] ··· io.json.ToJSON.mem_to_json_wide                                                                               ok
[ 54.55%] ··· ========= ==== ============= ============== =============== ==================
              --                                       frame
              --------- --------------------------------------------------------------------
                orient   df   df_date_idx   df_td_int_ts   df_int_floats   df_int_float_str
              ========= ==== ============= ============== =============== ==================
                split    0         0             0               0                0
               columns   0         0             0               0                0
                index    0         0             0               0                0
                values   0         0             0               0                0
               records   0         0             0               0                0
              ========= ==== ============= ============== =============== ==================

[ 56.82%] ··· io.json.ToJSON.time_to_json                                                                                   ok
[ 56.82%] ··· ========= ========== ============= ============== =============== ==================
              --                                          frame
              --------- --------------------------------------------------------------------------
                orient      df      df_date_idx   df_td_int_ts   df_int_floats   df_int_float_str
              ========= ========== ============= ============== =============== ==================
                split    93.9±5ms     94.2±9ms      98.5±6ms        102±5ms          102±20ms
               columns   95.6±2ms     308±20ms      352±40ms        321±60ms         313±40ms
                index    121±3ms      319±10ms      318±20ms        318±5ms          316±3ms
                values   78.3±3ms     78.6±1ms      79.3±1ms        88.4±1ms         87.5±1ms
               records   89.1±6ms     92.0±2ms      109±2ms         117±2ms          116±3ms
              ========= ========== ============= ============== =============== ==================

[ 59.09%] ··· io.json.ToJSON.time_to_json_wide                                                                              ok
[ 59.09%] ··· ========= ========== ============= ============== =============== ==================
              --                                          frame
              --------- --------------------------------------------------------------------------
                orient      df      df_date_idx   df_td_int_ts   df_int_floats   df_int_float_str
              ========= ========== ============= ============== =============== ==================
                split    110±1ms      110±2ms       168±4ms         154±6ms          181±20ms
               columns   123±1ms      139±2ms       190±2ms         176±2ms          200±3ms
                index    125±3ms      125±2ms      188±0.9ms        168±4ms          207±7ms
                values   109±9ms      112±7ms       170±10ms        149±10ms         191±20ms
               records   139±10ms     137±10ms      199±10ms        171±20ms         195±6ms
              ========= ========== ============= ============== =============== ==================

[ 61.36%] ··· io.json.ToJSONLines.time_delta_int_tstamp_lines                                                          147±3ms
[ 63.64%] ··· io.json.ToJSONLines.time_float_int_lines                                                                 163±4ms
[ 65.91%] ··· io.json.ToJSONLines.time_float_int_str_lines                                                            168±10ms
[ 68.18%] ··· io.json.ToJSONLines.time_floats_with_dt_index_lines                                                      131±7ms
[ 70.45%] ··· io.json.ToJSONLines.time_floats_with_int_idex_lines                                                      126±3ms
[ 72.73%] ··· Setting up io/json.py:200                                                                                     ok
[ 72.73%] ··· io.json.ToJSONMem.peakmem_float                                                                            62.3M
[ 75.00%] ··· io.json.ToJSONMem.peakmem_int                                                                              62.5M.
[ 75.00%] · For pandas commit 51db82d9 <master> (round 2/2):
[ 75.00%] ·· Building for conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 75.00%] ·· Benchmarking conda-py3.6-Cython-matplotlib-numexpr-numpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 77.27%] ··· io.json.ToJSON.mem_to_json                                                                                    ok
[ 77.27%] ··· ========= ==== ============= ============== =============== ==================
              --                                       frame
              --------- --------------------------------------------------------------------
                orient   df   df_date_idx   df_td_int_ts   df_int_floats   df_int_float_str
              ========= ==== ============= ============== =============== ==================
                split    0         0             0               0                0
               columns   0         0             0               0                0
                index    0         0             0               0                0
                values   0         0             0               0                0
               records   0         0             0               0                0
              ========= ==== ============= ============== =============== ==================

[ 79.55%] ··· io.json.ToJSON.mem_to_json_wide                                                                               ok
[ 79.55%] ··· ========= ==== ============= ============== =============== ==================
              --                                       frame
              --------- --------------------------------------------------------------------
                orient   df   df_date_idx   df_td_int_ts   df_int_floats   df_int_float_str
              ========= ==== ============= ============== =============== ==================
                split    0         0             0               0                0
               columns   0         0             0               0                0
                index    0         0             0               0                0
                values   0         0             0               0                0
               records   0         0             0               0                0
              ========= ==== ============= ============== =============== ==================

[ 81.82%] ··· io.json.ToJSON.time_to_json                                                                                   ok
[ 81.82%] ··· ========= =========== ============= ============== =============== ==================
              --                                           frame
              --------- ---------------------------------------------------------------------------
                orient       df      df_date_idx   df_td_int_ts   df_int_floats   df_int_float_str
              ========= =========== ============= ============== =============== ==================
                split    90.9±10ms     88.8±6ms      92.2±4ms        98.5±7ms         107±2ms
               columns    107±10ms     313±20ms      298±20ms        320±7ms          324±20ms
                index    104±0.8ms     309±10ms      308±10ms        311±5ms          343±10ms
                values    75.1±1ms     76.5±2ms      80.3±3ms        122±20ms         85.3±6ms
               records   95.0±10ms    92.1±10ms      123±10ms        125±7ms          141±30ms
              ========= =========== ============= ============== =============== ==================

[ 84.09%] ··· io.json.ToJSON.time_to_json_wide                                                                              ok
[ 84.09%] ··· ========= =========== ============= ============== =============== ==================
              --                                           frame
              --------- ---------------------------------------------------------------------------
                orient       df      df_date_idx   df_td_int_ts   df_int_floats   df_int_float_str
              ========= =========== ============= ============== =============== ==================
                split     112±6ms      111±5ms       214±30ms        184±10ms         189±30ms
               columns    126±3ms      140±2ms       205±10ms        179±4ms          211±20ms
                index     125±2ms      123±2ms       186±7ms         169±5ms          215±10ms
                values    113±10ms     107±2ms       163±6ms         146±1ms          174±3ms
               records   122±0.7ms     135±10ms      239±50ms        167±3ms          207±20ms
              ========= =========== ============= ============== =============== ==================

[ 86.36%] ··· io.json.ToJSONLines.time_delta_int_tstamp_lines                                                          147±2ms
[ 88.64%] ··· io.json.ToJSONLines.time_float_int_lines                                                                 164±4ms
[ 90.91%] ··· io.json.ToJSONLines.time_float_int_str_lines                                                             171±5ms
[ 93.18%] ··· io.json.ToJSONLines.time_floats_with_dt_index_lines                                                      123±6ms
[ 95.45%] ··· io.json.ToJSONLines.time_floats_with_int_idex_lines                                                     149±20ms
[ 97.73%] ··· Setting up io/json.py:200                                                                                     ok
[ 97.73%] ··· io.json.ToJSONMem.peakmem_float                                                                            62.3M
[100.00%] ··· io.json.ToJSONMem.peakmem_int                                                                              62.3M
.
BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

</details>

Anyone have thoughts on this?

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1647,3 +1647,25 @@ def test_tuple_labels(self, orient, expected):
         assert result == expected
+
+    @pytest.mark.parametrize("indent", [1, 2, 4])
+    def test_to_json_indent(self, indent):
```
i would add indent for some existing tests as this doesn’t test all of the code paths 

eg for different types i don’t think u r fully testing 

pick a value like 4 and use it consistently

WillAyd replied:
Sounds good. I'm thinking adding something for each orient will suffice, though if you are thinking more beyond that lmk

jreback replied:
looks good.

WillAyd committed:
Added tests for all orients

WillAyd committed:
merge conflict fixup

WillAyd committed:
Fixed table schema

WillAyd committed:
Simplified logic for building table schema

WillAyd committed:
Fixed test, removed breakpoint

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
Added whatsnew for fixed issue

WillAyd commented on pandas/io/json/_json.py:
```diff
@@ -315,19 +335,20 @@ def _write(
             default_handler,
+            indent,
         )
-        serialized = '{{"schema": {schema}, "data": {data}}}'.format(
```
Previously this was creating a formatted string from individual dumps of the JSON data. To get indent to work with that formatted string adds a lot of code, so instead I put everything into an OrderedDict and am letting the JSON encoder do the heavy lifting.

This fixed a bug as indicated in the whatsnew but also shed some light on #28256 which I think is best tackled separately.

**Note** that this also changes the whitespace in use for the table format, which was inconsistent before anyway. The JSON encoder doesn't put a space after each colon whereas this formatted string did. Whitespace is insignificant in JSON so I don't think this really needs a whatsnew mention but can be convinced otherwise

WillAyd reviewed (commented):
Expanded test coverage for all formats. Uncovered a few inconsistencies with datetimelike handling in the table writer which I've either addressed directly or opened a follow up issue for

WillAyd committed:
Ran clang-format on objToJSON.c

WillAyd committed:
lint fixups

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
whitespace fixup

WillAyd committed:
Py35 compat

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
isort and lint fixups

WillAyd commented:
Couple 35 compat things came up when simplifying table build. Not necessarily relevant as order of JSON objects is undefined and 35 will be dropped by time this goes live anyway, but added appropriate compat to get tests to pass. 

In any case ready for review again

jreback commented on pandas/core/generic.py:
```diff
@@ -2342,6 +2344,11 @@ def to_json(
+        indent : integer, default 0
+           Length of whitespace used to indent each record.
+
+           .. versionadded:: 1.0.0
```
does this match the semantics of the stdlib, e.g. default is ``None`` and 0 means insert newlines, also can be a string, is that supported? https://docs.python.org/3/library/json.html

WillAyd replied:
No this only accepts int for now and 0 does not insert new lines. This matches ujson behavior instead of stdlib

WillAyd replied:
It should be possible to support this in the long term I just think will take a little bit of effort to bridge the gap with our vendored ujson. Thinking for now it might make the most sense to change the signature to `indent=None` and have that map to `indent=0`, documenting the the `indent=0` behavior is different from stdlib. 

I think that should avoid a deprecation cycle in the future for `indent=0` behavior if we eventually can mirror the stdlib's support

jreback commented on pandas/tests/io/json/test_json_table_schema.py:
```diff
@@ -234,7 +236,11 @@ def test_build_series(self):
         )
-        assert result == expected
+
+        if PY35:
```
can you encapsulate this in a function, assert_results_equal(result, expected) (do it in this file i think)

jreback replied:
can you do this here

WillAyd replied:
Sorry missed this before but should be done now

jreback commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1647,3 +1647,152 @@ def test_tuple_labels(self, orient, expected):
+            ),
+            (
+                "columns",
+                """{
```
what about invalid values for indent, e.g. < 0 or a non-integer?

WillAyd replied:
None integers raise a TypeError. I'll add the validation of non-negative

WillAyd committed:
Changed default indent to None

WillAyd committed:
Validate int input

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd commented:
Validation added and changed the default to `indent=None`, with a documentation note on `index=0` behavior and how it differs from stdlib

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
Added helper function for Py35 compat

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
Reverted line removal

WillAyd commented:
Ping on this one - I think good to go

gfyoung commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -191,6 +191,7 @@ I/O
 - Bug in :meth:`DataFrame.to_json` where using a Tuple as a column or index value and using ``orient="columns"`` or ``orient="index"`` would produce invalid JSON (:issue:`20500`)
 - Improve infinity parsing. :meth:`read_csv` now interprets ``Infinity``, ``+Infinity``, ``-Infinity`` as floating point values (:issue:`10065`)
 - Bug in :meth:`DataFrame.to_csv` where values were truncated when the length of ``na_rep`` was shorter than the text input data. (:issue:`25099`)
+- Bug in :meth:`DataFrame.to_json` where a datetime column label would not be written out in iso format with ``orient="table"`` (:issue:`28130`)
```
```suggestion
- Bug in :meth:`DataFrame.to_json` where a datetime column label would not be written out in ISO format with ``orient="table"`` (:issue:`28130`)
```

gfyoung commented on pandas/_libs/src/ujson/lib/ultrajsonenc.c:
```diff
@@ -960,24 +974,28 @@ void encode(JSOBJ obj, JSONObjectEncoder *enc, const char *name,
             enc->iterEnd(obj, &tc);
-            Buffer_Reserve(enc, 2);
+            Buffer_AppendIndentNewlineUnchecked (enc);
+            Buffer_AppendIndentUnchecked (enc, enc->level);
```
Is there ever a time when the second argument doesn't come from `enc` ?

Perhaps we can just collapse to one argument.

WillAyd replied:
Well spotted and certainly logical, but this is just a straight vendor from ujson, so I think better to limit customization

gfyoung replied:
Fair enough.  Can we note that for those functions for future selves so that we don't accidentally refactor?

WillAyd replied:
Sure - just added

gfyoung commented on pandas/core/generic.py:
```diff
@@ -2249,17 +2250,18 @@ def to_excel(
+        compression: Optional[str] = "infer",
+        index: bool_t = True,
+        indent: Optional[int] = None,
+    ) -> Optional[str]:
```
Nit [no action needed]: I know that we've been pushing typing, but in the future, it would be preferable to separate out this effort from other changes like this enhancement.  Makes things easier to review.

WillAyd replied:
Yea for sure - these do get kind of complicated when you annotate a new module altogether. Hopefully less of an issue going forward especially with all of the work @simonjayhawkins has been putting in

gfyoung commented on pandas/core/generic.py:
```diff
@@ -2399,6 +2413,13 @@ def to_json(
+        if indent is None:
+            int_indent = 0
+        else:
+            int_indent = indent
```
Can't we collapse this to:

~~~python
indent = indent or 0
~~~

BTW, at this point, we don't need another variable because the `config` call already confirms an `int` or `None`.

WillAyd replied:
It's valid Python but doesn't jive with typing, because `indent` local to this function is `Optional[int]` but everywhere else it is just `int`. This is an easy way to limit the types for the checker, since `int_indent` will always be `int`

WillAyd replied:
There is a flag that supports this:

https://github.com/python/mypy/pull/6197

But not enabled, assuming we even wanted that

WillAyd replied:
Actually I stand corrected - looks like it can narrow the type here implicitly. Nice catch!

gfyoung commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -1647,3 +1647,156 @@ def test_tuple_labels(self, orient, expected):
+        if PY35:
+            assert json.loads(result) == json.loads(expected)
+        else:
+            assert result == expected
```
Why do we need this again exactly?

WillAyd replied:
Comes back to dict ordering - cannot be guaranteed in Py35. Ultimately we'll be dropping Py35 by the time this goes live so somewhat of a moot point, but for now the compat is an easy indication of what needs to be cleaned up when we officially drop 35

WillAyd replied:
This comment might help explain more too https://github.com/pandas-dev/pandas/pull/28130/files#r320046798

gfyoung replied:
Makes sense.

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into json-indent2

WillAyd committed:
Whatsnew fix

WillAyd committed:
Simplified indent code

WillAyd committed:
Added comment for func

WillAyd committed:
whitespace doc fixup

WillAyd commented:
Thanks for the feedback @gfyoung . If no other comments I plan to merge this on Wednesday

botzill commented:
Hi.

Is indent feature still in pending?
